synapse/cores/common.py

```diff
@@ -2619,6 +2622,109 @@ def _popCoreXact(self):
     def _getCoreXact(self, size):
         raise NoSuchImpl(name='_getCoreXact')
 
+    def _onAddSynTagDark(self, mesg):
+        tufo = mesg[1].get('tufo')  # type: tuple(str, dict)
+        tag = mesg[1].get('tag')  # type: str
+        self.addTufoDark(tufo, 'tag', tag)
+
+    def _onDelSynTagDark(self, mesg):
+        tufo = mesg[1].get('tufo')  #type: ((str, dict))
+        tag = mesg[1].get('tag')
+        self.delTufoDark(tufo, 'tag', tag)
+
```

Comment:
These APIs are *great*.  You've brought up an interesting question i hadn't really thought of.  I think we should consider "dark rows" to always be "cortex local" ( and if consistency is needed, they should be created by other sync actions, as would happen with the tags etc ) and as such (being local) they should not need to fire events etc.  The APIs are a nice way to get the rows added, but if some module wants to do black magic ( like i'm about to suggest for tag performance ;) ) i think bypassing the dark API all together should be considered acceptable.

So, using the onAddSynTagDark() / onDelSynTagDark() is definitely the most elegant way to do this. :)  However...  for tag perf's sake, i'd like to get the dark rows in with the addRows() call that is already in addTufoTag to minimize the number of round trips through the storage APIs.  It might get a bit down and dirty, but i think it's worth it, especially for add.